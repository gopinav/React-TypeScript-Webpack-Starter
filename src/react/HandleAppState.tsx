import { Button, IconButton } from "@material-ui/core";
import Snackbar from "@material-ui/core/Snackbar";
import CloseIcon from "@material-ui/icons/Close";
import MuiAlert, { AlertProps } from "@material-ui/lab/Alert";
import {Fragment, useState} from "react";
import { Workbox } from "workbox-window";

export const HandleAppState = ({wb}: { wb: Workbox; }) =>
{
    const strings = 
    {
        UPDATE: "New Update Available ðŸŽ‰",
        ONLINE: "Back on line.",
        OFFLINE: "Internet connection lost."
    };
     
    const [workerWaiting, setWorkerWaiting] = useState(false);
    const [toggleOnline, setToggleOnline] = useState({online: true, change: false});

    wb.addEventListener("waiting", () => setWorkerWaiting(true));

    window.addEventListener("online", () => setToggleOnline({online: true, change: true}));
    window.addEventListener("offline", () => setToggleOnline({online: false, change: true}));

    function Alert(props: AlertProps) 
    {
        return (
            <MuiAlert
                elevation={6} variant="filled" {...props}
                onClose={() => setToggleOnline({online: toggleOnline.online, change: false})} 
            />
        );
    }

    function updateWorker() 
    {
        setWorkerWaiting(false);
        wb.messageSkipWaiting();
    }
    
    return(
        <>
            <Snackbar
                open={workerWaiting}
                autoHideDuration={6000}
                message={strings.UPDATE}
                anchorOrigin={{ vertical: "top", horizontal: "center" }}
                action={
                    <Fragment>
                        <Button color="secondary" size="small" 
                            onClick={updateWorker}> 
                            UPDATE 
                        </Button>
                        <IconButton size="small" aria-label="close" onClick={() => setWorkerWaiting(false)}>
                            <CloseIcon fontSize="small" />
                        </IconButton>
                    </Fragment>
                }
            />
            <Snackbar open={toggleOnline.change} anchorOrigin={{ vertical: "top", horizontal: "center" }}>
                <Alert severity={toggleOnline.online ? "success": "error"} >
                    {toggleOnline.online ? strings.ONLINE : strings.OFFLINE}
                </Alert>
            </Snackbar>
        </>
    );
};
