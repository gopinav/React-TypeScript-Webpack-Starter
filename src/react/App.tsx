import { HandleAppState } from "./HandleAppState";
import { Counter } from "./Counter";
import { Workbox } from "workbox-window";

import logo from "../assets/icons/logo512.png";
import gif from "../assets/images/giphy.gif";

export const App = ({wb}: { wb: Workbox; }) => 
{
    return (
        <>
            <HandleAppState wb={wb} /> 
            <h1>React TypeScript Webpack Starter Template</h1>
            <Counter />
            <img src={logo.src} srcSet={logo.srcSet} width={logo.width} height={logo.height} loading="lazy" alt="React logo" />
            <img src={gif} loading="lazy" alt="Silicon valley gif" />
        </>
    );
};
