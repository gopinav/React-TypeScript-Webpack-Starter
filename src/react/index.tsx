import ReactDOM from "react-dom";
import { App } from "./App";
import { Workbox } from "workbox-window";

const wb = new Workbox("sw.js");

ReactDOM.render(<App wb={wb} />, document.getElementById("root"));

if ("serviceWorker" in navigator) 
    wb.register();
    
