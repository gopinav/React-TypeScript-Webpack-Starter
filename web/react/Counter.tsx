import { useState } from "react";

export const Counter = () => 
{
    const [count, setCount] = useState(0);
    return (
        <div>
            <button onClick={() => setCount((c) => c + 1)}>
                Count - {count}
            </button>
        </div>
    );
};
