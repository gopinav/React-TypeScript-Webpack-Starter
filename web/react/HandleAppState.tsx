import { Button, IconButton } from "@material-ui/core";
import Snackbar from "@material-ui/core/Snackbar";
import CloseIcon from "@material-ui/icons/Close";
import MuiAlert, { AlertProps } from "@material-ui/lab/Alert";
import {Fragment, useState} from "react";
import { Workbox } from "workbox-window";

export const HandleAppState = ({wb}: { wb: Workbox; }) =>
{
    const strings = 
    {
        UPDATE: "New Update Available ðŸŽ‰",
        ONLINE: "Back on line.",
        OFFLINE: "Internet connection lost."
    };
     
    const [workerWaiting, setWorkerWaiting] = useState(false);
    const [toggleOnline, setToggleOnline] = useState({online: navigator.onLine, change: !navigator.onLine});

    wb.addEventListener("waiting", () => setWorkerWaiting(true));

    window.addEventListener("online", () => setToggleOnline({online: true, change: true}));
    window.addEventListener("offline", () => setToggleOnline({online: false, change: true}));

    function Alert(props: AlertProps) 
    {
        return (
            <MuiAlert
                elevation={6} variant="filled" {...props}
                onClose={() => setToggleOnline({online: toggleOnline.online, change: false})} 
            />
        );
    }

    function updateWorker() 
    {
        wb.addEventListener("controlling", () => window.location.reload());
        setWorkerWaiting(false);
        wb.messageSkipWaiting();
    }
    
    return(
        <>
            <Snackbar
                open={workerWaiting}
                autoHideDuration={60000}
                message={strings.UPDATE}
                anchorOrigin={{ vertical: "top", horizontal: "center" }}
                action={
                    <Fragment>
                        <Button color="secondary" size="small" 
                            onClick={updateWorker}> 
                            UPDATE 
                        </Button>
                        <IconButton size="small" aria-label="close" onClick={() => setWorkerWaiting(false)}>
                            <CloseIcon fontSize="small" />
                        </IconButton>
                    </Fragment>
                }
                onClose={() => setWorkerWaiting(false)}
            />
            <Snackbar open={toggleOnline.change} 
                anchorOrigin={{ vertical: "top", horizontal: "center" }}
                autoHideDuration={10000}
                onClose={() => setToggleOnline({online: navigator.onLine, change: false})}
            >
                <Alert severity={toggleOnline.online ? "success": "error"} >
                    {toggleOnline.online ? strings.ONLINE : strings.OFFLINE}
                </Alert>
            </Snackbar>
        </>
    );
};
