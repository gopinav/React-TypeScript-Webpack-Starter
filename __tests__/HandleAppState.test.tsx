/**
 * @jest-environment jsdom
 */

import React from "react";
import { mount, shallow } from "enzyme";
import { HandleAppState } from "../web/react/HandleAppState";
import { act } from "react-dom/test-utils";

const snackbar = "WithStyles(ForwardRef(Snackbar))"; 


describe("HandleAppState Component", () => 
{
    let wb;

    beforeEach(() => 
    {
        wb = {
            addEventListener: (prop: string | number, callback: () => void) => wb[prop] = callback,
            messageSkipWaiting: jest.fn(),
            waiting: () => null,
            controlling: () => null 
        };
    });

    afterEach(() => 
    {
        jest.useRealTimers();
        jest.restoreAllMocks();
    });

    it("should render correctly by default", () => expect(shallow(<HandleAppState wb={wb} />)).toMatchSnapshot() );

    describe("Should handle service worker states", ()=>
    {
        it("Update sw and hide snackbar on button click.", () => 
        {
            const warper = mount(<HandleAppState wb={wb} />);

            act(() => wb.waiting());
            warper.update();

            expect(warper.children(snackbar).first().prop("open")).toBe(true);

            warper.find("button").first().simulate("click");
        
            expect(wb.messageSkipWaiting).toBeCalled();
            expect(warper.children(snackbar).first().prop("open")).toBe(false);

            warper.unmount();
        });

        it("Close without updating after a minute.", () => 
        {
            jest.useFakeTimers();

            const warper = mount(<HandleAppState wb={wb} />);

            act(wb.waiting);
            warper.update();

            expect(warper.children(snackbar).first().prop("open")).toBe(true);

            act(() => (jest.advanceTimersByTime(60 * 1000), undefined));
            warper.update();
        
            expect(wb.messageSkipWaiting).toHaveBeenCalledTimes(0);
            expect(warper.children(snackbar).first().prop("open")).toBe(false);

            warper.unmount();
        });

    });

    describe("Should handle connectivity states", ()=> 
    {
        let functions = {};

        beforeEach(() => 
        {
            functions = {};
            jest.spyOn(window, "addEventListener").mockImplementation((type, fn) => (functions[type] = fn));
        });

        function setOnlineState(isOnline: boolean) 
        {
            jest.spyOn(navigator, "onLine", "get").mockReturnValue(isOnline);

            const toCall = (isOnline ? functions["online"] : functions["offline"]);

            if(toCall) act(toCall);
        }

        it("Should show snackbar when loading in offline mode.", () => 
        {
            setOnlineState(false);
            expect(shallow(<HandleAppState wb={wb} />)).toMatchSnapshot();
        });

        it("Should hide offline 10 seconds after loading in offline mode", () => 
        {
            jest.useFakeTimers();
            setOnlineState(false);

            const warper = mount(<HandleAppState wb={wb} />);

            expect(warper.children(snackbar).at(1).prop("open")).toBe(true);

            act(() => (jest.advanceTimersByTime(10 * 1000), undefined));
            warper.update();

            expect(warper.children(snackbar).at(1).prop("open")).toBe(false);

            warper.unmount();
        });

        it("Should toggle online and offline states correctly", () => 
        {
            const warper = shallow(<HandleAppState wb={wb} />);

            setOnlineState(false);
            warper.update();

            expect(warper.children(snackbar).at(1).prop("open")).toBe(true);
            expect(warper.find("Alert").first().prop("severity")).toBe("error");

            setOnlineState(true);
            warper.update();

            expect(warper.children(snackbar).at(1).prop("open")).toBe(true);
            expect(warper.find("Alert").first().prop("severity")).toBe("success");

            warper.unmount();
        });
    });

});
