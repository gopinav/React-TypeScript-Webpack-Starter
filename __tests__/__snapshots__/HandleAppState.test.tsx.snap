// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandleAppState Component Should handle connectivity states Should show snackbar when loading in offline mode. 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Snackbar))
    action={
      <React.Fragment>
        <WithStyles(ForwardRef(Button))
          color="secondary"
          onClick={[Function]}
          size="small"
        >
          UPDATE
        </WithStyles(ForwardRef(Button))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="close"
          onClick={[Function]}
          size="small"
        >
          <Memo
            fontSize="small"
          />
        </WithStyles(ForwardRef(IconButton))>
      </React.Fragment>
    }
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "top",
      }
    }
    autoHideDuration={60000}
    message="New Update Available ðŸŽ‰"
    onClose={[Function]}
    open={false}
  />
  <WithStyles(ForwardRef(Snackbar))
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "top",
      }
    }
    autoHideDuration={10000}
    onClose={[Function]}
    open={true}
  >
    <Alert
      severity="error"
    >
      Internet connection lost.
    </Alert>
  </WithStyles(ForwardRef(Snackbar))>
</Fragment>
`;

exports[`HandleAppState Component should render correctly by default 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Snackbar))
    action={
      <React.Fragment>
        <WithStyles(ForwardRef(Button))
          color="secondary"
          onClick={[Function]}
          size="small"
        >
          UPDATE
        </WithStyles(ForwardRef(Button))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="close"
          onClick={[Function]}
          size="small"
        >
          <Memo
            fontSize="small"
          />
        </WithStyles(ForwardRef(IconButton))>
      </React.Fragment>
    }
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "top",
      }
    }
    autoHideDuration={60000}
    message="New Update Available ðŸŽ‰"
    onClose={[Function]}
    open={false}
  />
  <WithStyles(ForwardRef(Snackbar))
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "top",
      }
    }
    autoHideDuration={10000}
    onClose={[Function]}
    open={false}
  >
    <Alert
      severity="success"
    >
      Back on line.
    </Alert>
  </WithStyles(ForwardRef(Snackbar))>
</Fragment>
`;
